keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,531 @@
+
+Similar features for protecting PV and PVC already exist as [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md).
+
+
+## Motivation
+
+This feature aims to protect Secrets/ConfigMaps from being deleted while they are in-use.
+Secrets can be used by below ways:
+- From Pod:
+  - [Mounted as data volumes](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables)
+  - [Generic ephemeral volumes
+](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes) (can be handled as CSI PV below)
+- From PV:
+  - [CSI](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html):
+    - provisioner secret
+    - controller publish secret
+    - node stage secret
+    - node publish secret
+    - controller expand secret
+  - non-CSI:
+    - dependent on each storage driver and will be deprecated soon (Out of scope)
+- [From VolumeSnapshot](https://kubernetes-csi.github.io/docs/secrets-and-credentials-volume-snapshot-class.html):
+  - snapshotter secret
+
+ConfigMaps can be used by below ways:
+- From Pod:
+  - [Mounted as files](https://kubernetes.io/docs/concepts/configuration/configmap/#using-configmaps-as-files-from-a-pod)
+  - [Exposed as environment variables](https://kubernetes.io/docs/concepts/configuration/configmap/#configmaps-and-pods)
+
+### Goals
+
+- Protect Secrets/ConfigMaps from being deleted while they are in use.
+
+### Non-Goals
+
+- Protect important Secrets/ConfigMaps that aren't in use from being deleted
+- Protect resources other than Secrets/ConfigMaps from being deleted.
+
+## Proposal
+
+New controllers to protect Secrets/ConfigMaps are introduced.
+
+### User Stories
+
+#### Story 1
+
+A user creates Secrets/ConfigMaps and a pod using them. Then, the user mistakenly delete them while the pod is using them.
+They are protected until the pod using them is deleted.
+
+#### Story 2
+
+A user creates a volume that uses a certain secret in the same namespace. Then, the user delete the namespace.
+The secret is protected until the volume using the secret is deleted and the deletion of the volume succeeds.
+
+#### Story 3
+
+A user really would like to delete a secret despite that it is used by other resources.
+The user force delete the secret while it is used by other resources, and the secret isn't protected and is actually deleted.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Compatibility:
+  - There might be many existing scripts that don't care the order of deletion. Therefore, such scripts might stuck on Secret/ConfigMap deletion, if the deletion of the resources using the Secrets/ConfigMaps are done later.
```

Comment:
Good question. Current protections are done for only below direction, therefore ABBA deadlock style of issue won't happen. One thing that I'm worrying about in the future is the relationship with CSI secret store.
(I'm thinking about that it eventually should be wrapped by Secret to be protected in the same way, however, by doing so, it will make cyclic dependency). I will consider adding this kind of discussions to the KEP.

[pvc-protection]
Pod -> PVC

[pv-protection]
PVC -> PV

[this KEP]
Pod -> Secret
PV -> Secret
PVC -> Secret
PVC -> SC -> Secret